sdk/core/azure-core/azure/core/paging.py

```diff
@@ -23,86 +23,93 @@
-    :type response: ~azure.core.pipeline.transport.HttpResponse
-    :param callable command: Function to retrieve the next page of items.
-    :param Deserializer deserializer: a Deserializer instance to use
-    """
-    _validation = {}  # type: Dict[str, Dict[str, Any]]
-    _attribute_map = {}  # type: Dict[str, Dict[str, Any]]
-
-    def __init__(self, command, deserializer, **kwargs):
-        # type: (Callable[[str], HttpResponse], Deserializer, Any) -> None
-        super(Paged, self).__init__(**kwargs)  # type: ignore
-        # Sets next_link, current_page, and _current_page_iter_index.
-        self.next_link = ""
-        self.current_page = []  # type: List[Model]
-        self._current_page_iter_index = 0
-        self._deserializer = deserializer
-        self._get_next = command
-        self._response = None  # type: Optional[HttpResponse]
+import itertools
+from typing import (
+    Dict,
+    Any,
+    List,
+    Callable,
+    Optional,
+    TypeVar,
+    Iterator,
+    Iterable,
+    Tuple,
+    Type,
+)  # pylint: disable=unused-import
+import logging
+
+
+_LOGGER = logging.getLogger(__name__)
+
+ReturnType = TypeVar("ReturnType")
+ResponseType = TypeVar("ResponseType")
+
+
+class PageIterator(Iterator[Iterator[ReturnType]]):
+    def __init__(self, get_next, extract_data, continuation_token=None):
+        # type: (Callable[[str], ResponseType], Callable[[ResponseType], Tuple[str, Iterable[ReturnType]]], Optional[str]) -> None
+        """Return an iterator of pages.
+
+        :param get_next: Callable that take the continuation token and return a HTTP response
+        :param extract_data: Callable that take an HTTP response and return a tuple continuation token,
+         list of ReturnType
+        :param str continuation_token: The continuation token needed by get_next
+        """
+        self._get_next = get_next
+        self._extract_data = extract_data
+        self.continuation_token = continuation_token
+        self._did_a_call_already = False
+        self._response = None
+        self._current_page = None
 
     def __iter__(self):
         """Return 'self'."""
-        # Since iteration mutates this object, consider it an iterator in-and-of
-        # itself.
         return self
 
-    @classmethod
-    def _get_subtype_map(cls):
-        """Required for parity to Model object for deserialization."""
-        return {}
+    def __next__(self):
+        # type: () -> Iterator[ReturnType]
+        if self.continuation_token is None and self._did_a_call_already:
+            raise StopIteration("End of paging")
 
-    def _advance_page(self):
-        # type: () -> List[Model]
-        """Force moving the cursor to the next azure call.
+        self._response = self._get_next(self.continuation_token)
+        self._did_a_call_already = True
+
+        self.continuation_token, self._current_page = self._extract_data(self._response)
+        if not self._current_page:
```

Comment:
Would it be better to have `self._extract_data` raise an exception and clean up the `continuation_token` and `_current_page` in the exception handler and raise a `StopIteration` exception?

Maybe this is just an overreaction to some of the previous self-modifying methods that we had - to have the assignment of `self._current_page` from the return value of `self._extract_data`  rather from within the method is probably enough. But just throwing it out there to see what people think...

Reply:
The idea was to avoid as much as possible for the callbacks to know anything about the paging implementation. If the page received evaluates to false (likely an empty list), then assume the paging is done.
It's because we have two kinds of conditions:
- continuation_token is None: please process the current, but do not do a future call
- current_page is "false": the last call actually didn't return element, stop now

in ARM, only the first one happens. In Storage, I found situation where the stopping criteria was "I received an empty list"

Reply:
At least for KeyVault, returning an empty list is not an indicator that you are done paging. For storage, it (a page is empty) can happen on the server side as well, but I don't know if they are responding with an empty list for a client in situations other than them trying to end the iteration...

Reply:
Do you want me to change this to 
``` python
if not self.continuation_token and not self._current_page
```
in case there is a continuation token, but an empty list? ( I can't imagine that, but still if you say it's possible :))

Reply:
Yes. KeyVault will be broken otherwise.

Reply:
Fixed